keps/sig-node/4010-head-request-probes/README.md

```diff
@@ -0,0 +1,265 @@
+        value: some-value
+    queryParams:
+      - name: token
+        value: abc123
+    path: /health
+    port: 8080
+    scheme: HTTPS
+    timeoutSeconds: 3
+    periodSeconds: 5
+```
+
+In this example, the readiness probe is configured to use an HTTP HEAD request with custom headers, query parameters, and a specific path, port, and scheme.
+
+There is an [untested solution how this feature might be implemented](https://github.com/kubernetes/kubernetes/commit/e3d556f0920ff736fec5df73c42f6db93f436a37).  
+Note: this solution is also missing a feature gate, which will be required for the final implementation.
+
+### Risks and Mitigations
+
+Introducing support for HTTP HEAD requests in probes may pose the following risks:
+
+- Increased complexity: Adding support for HEAD requests may introduce additional complexity to the Kubernetes codebase. This risk can be mitigated through careful design and thorough testing.
+- Backward compatibility: Existing readiness and liveness probes that use GET requests may need to be updated to utilize the new HEAD functionality. Documentation and upgrade guides can help users transition smoothly.
+
+## Design Details
+
+1. Extend the existing readiness and liveness probe specifications to include an optional `httpHead` field.
+2. When `httpHead` is specified, the probe will use an HTTP HEAD request instead of the default GET request.
+3. The `httpHead` field will accept an object that allows specifying additional headers, query parameters, and any other relevant fields required for the HEAD request.
+4. If `httpHead` is not specified, the probe will default to using a GET request, ensuring backward compatibility with existing probes.
+5. Modify the Kubernetes health check mechanism to handle HEAD requests appropriately, including parsing the response and determining the pod's health based on the HEAD response status code.
+
+### Test Plan
+
+The following tests should be performed to ensure the proper functioning of the new HEAD request support:
+
+#### Prerequisite testing updates
+
+- Update existing tests for readiness and liveness probes to include scenarios using HEAD requests.
+
+#### Unit tests
+
+- Add unit tests to verify the correct handling of HEAD requests in the Kubernetes API server and Kubelet components.
+- Test the behavior of the probes when HEAD requests are used.
+
+#### Integration tests
+
+- Perform integration tests to validate the interoperability of the updated components with other Kubernetes functionality.
+- Verify the correct behavior of HEAD-based probes in various scenarios, including failures and recoveries.
+
+#### e2e tests
+
+- Create end-to-end tests that cover the entire lifecycle of pods with HEAD-based probes.
+- Validate the behavior of probes in real-world scenarios, ensuring they function as expected.
+
+### Graduation Criteria
+
+The support for HEAD requests in readiness and liveness probes will undergo the following graduation stages:
```

Comment:
criteria typically include level of testing and level of adoption we expect to promote the feature between stages

Reply:
Updated the graduation criteria according analogues KEP 3695